### PR TITLE
Rename test function for protocol proxy consistency

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Rename test function for protocol proxy consistency
- Implements protocol-mode byte-stream proxy loops for container IO. Adds a dedicated protocol proxy path (stdin -> container stdin, container stdout -> host stdout, container stderr -> host stderr) and wires ExecMode::Protocol to a new proxy module. Includes unit and behavioural tests, as well as documentation updates. This is a runtime implementation, not just planning scaffolding.

## Changes
- Added new protocol proxy module: `src/engine/connection/exec/protocol.rs` implementing a dedicated protocol proxy with host IO seams and byte-preserving forwarding.
- Wired protocol mode into the engine:
  - Updated `src/engine/connection/exec/mod.rs` to dispatch `ExecMode::Protocol` to the new protocol runner.
- Tests and behavioural coverage:
  - Added unit helpers under `src/engine/connection/exec/tests/proxy_helpers.rs`.
  - Added protocol behavioural tests under `src/engine/connection/exec/tests/protocol_proxy_bdd.rs`.
  - Added protocol proxy feature and step definitions under `tests/features/protocol_proxy.feature` and `tests/features` folder.
- Documentation and planning artifacts:
  - Added new ExecPlan doc: `docs/execplans/2-5-2-byte-stream proxy loops.md` detailing goals, constraints, risks, and staged plan.
  - Minor formatting improvements to `docs/execplans/1-2-2-eyre-report.md` for consistency.
- Repository context pack: added placeholder lock at `.agents/mcp/context_pack/packs/.repo.lock`.
- Roadmap and docs:
  - Updated `docs/podbot-design.md` to reflect protocol-mode proxying decisions.
  - Marked Step 2.5.2 as complete in `docs/podbot-roadmap.md`.

## Rationale
- Isolates protocol-mode IO from interactive attached-mode semantics (no local terminal echo, no resize handling in protocol mode, tty=false).
- Establishes a testable host-IO seam for proxy loops with injected readers/writers, enabling unit and behavioural tests before full CLI integration.
- Provides a concrete, maintainable path for implementing the three proxy loops and wiring into ExecMode::Protocol, aligning with the Step 2.5.2 roadmap.

## Validation and Acceptance
- [x] Stage A: Split protocol proxying from interactive terminal attachment (protocol module separated from attached path).
- [x] Stage B: Introduce a testable host-IO seam for proxy loops (IO abstraction for protocol proxy).
- [x] Stage C: Implement protocol byte-stream loops and shutdown ordering (stdin -> container stdin; container stdout -> host stdout; container stderr -> host stderr).
- [x] Stage D: Wire protocol-mode exec orchestration and error mapping (ExecMode::Protocol dispatch wired).
- [x] Stage E: Unit tests with rstest covering forward, routing, and error paths.
- [x] Stage F: Behavioural tests with rstest-bdd for protocol proxying.
- [x] Stage G: Documentation updates and roadmap alignment.

Documentation and Roadmap impact
- The new ExecPlan doc codifies the goals and staged plan for Step 2.5.2.
- Podbot design and user docs updated to describe protocol-mode behaviour and its separation from interactive mode.
- Roadmap checkbox for Step 2.5.2 is now marked done; follow-ups will address further hardening (e.g., buffering policies) in later steps.

## Next steps
- Integrate the protocol proxy path with the host workflow and ExecMode::Protocol dispatch in all orchestration layers.
- Expand tests with additional edge cases (EOF handling, partial writes, and shutdown race conditions).
- Update user-facing docs to reflect the protocol-mode behaviour and limitations for embedders/hosted servers.

📎 **Task**: https://www.devboxer.com/task/0bf86d73-df1f-4c40-ae07-9548ab682652

## Summary by Sourcery

Implement dedicated protocol-mode byte-stream proxy loops for container exec and wire ExecMode::Protocol to a new proxy module, with accompanying tests and documentation updates.

Enhancements:
- Introduce a dedicated protocol exec proxy module that forwards stdin, stdout, and stderr as raw byte streams and ignores interactive-only behaviours like resize and stdin echo.
- Update engine exec dispatch so ExecMode::Protocol uses the new protocol proxy path instead of the interactive attached helper.

Documentation:
- Add an ExecPlan documenting Step 2.5.2 byte-stream proxy loops and refine existing exec plan formatting.
- Update podbot design and user guide docs to describe protocol-mode stdio proxy behaviour, echo suppression, and shutdown ordering.
- Mark roadmap Step 2.5.2 as complete after landing protocol proxy loops.

Tests:
- Add unit tests for protocol proxy helpers covering byte forwarding, routing between stdout/stderr, and error handling for IO and daemon stream failures.
- Add behavioural BDD tests and feature scenarios that exercise protocol exec byte proxying via injected host/container IO seams.

📎 **Task**: https://www.devboxer.com/task/3cdeba59-1976-4c10-bf66-1b603aaf87b8

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/engine/connection/exec/tests/proxy_helpers.rs](https://github.com/leynos/podbot/pull/69/files/4a48596d964e939e0836aa5fd46bae4ef67cd68d#diff-43c74a15cc94f67e05066650495d263bd03804e7fd59b63faa9ebe3513a3f7dd)

Comment on file

~~~rust
fn protocol_proxy_maps_daemon_stream_errors(runtime: RuntimeFixture) {
    let runtime_handle = runtime.expect("runtime fixture should initialize");
    let request = make_protocol_request().expect("protocol request should build");
    let host_stdin = runtime_handle
        .block_on(make_host_stdin(b""))
        .expect("host stdin should build");
    let output = make_output_stream(vec![Err(BollardError::RequestTimeoutError)]);

    let result = runtime_handle.block_on(run_protocol_session_with_io_async(
        &request,
        output,
        Box::pin(RecordingInputWriter::new()),
        ProtocolProxyIo::new(host_stdin, RecordingWriter::new(), RecordingWriter::new()),
    ));

    assert_exec_failed_message(result, "exec stream failed");
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=69&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fproxy_helpers.rs&method=)
The module contains 6 functions with similar structure: protocol_proxy_forwards_stdin_bytes_and_shutdowns_input,protocol_proxy_ignores_stdin_echo_chunks,protocol_proxy_maps_container_input_flush_failure,protocol_proxy_maps_daemon_stream_errors and 2 more functions

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/engine/connection/exec/tests/protocol_proxy_bdd.rs](https://github.com/leynos/podbot/pull/69/files/4a48596d964e939e0836aa5fd46bae4ef67cd68d#diff-53118f4c052a3bcdbe04e6d93188ee187c8f2e743cbf34e81ba1721a29557d9e)

Comment on file

~~~rust
fn host_stdout_receives(protocol_proxy_state: &ProtocolProxyState, text: String) -> StepResult<()> {
    let expected = normalize_feature_text(&text);
    let actual = protocol_proxy_state
        .host_stdout
        .get()
        .ok_or_else(|| String::from("host stdout should be recorded"))?;
    if actual == expected {
        Ok(())
    } else {
        Err(format!("expected host stdout {expected:?}, got {actual:?}"))
    }
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=69&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fprotocol_proxy_bdd.rs&method=)
The module contains 3 functions with similar structure: container_stdin_receives,host_stderr_receives,host_stdout_receives

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/engine/connection/exec/tests/proxy_helpers.rs](https://github.com/leynos/podbot/pull/69/files/ed7ec8a188ab254d83a447b0ed1d4c6629475307#diff-43c74a15cc94f67e05066650495d263bd03804e7fd59b63faa9ebe3513a3f7dd)

Comment on file

~~~rust
fn protocol_proxy_routes_stdout_and_console_to_host_stdout(runtime: RuntimeFixture) {
    let host_stdout = RecordingWriter::new();
    let host_stderr = RecordingWriter::new();
    let stdout_bytes = host_stdout.bytes.clone();
    let stderr_bytes = host_stderr.bytes.clone();
    let output = make_output_stream(vec![
        Ok(LogOutput::StdOut {
            message: Vec::from(&b"alpha"[..]).into(),
        }),
        Ok(LogOutput::Console {
            message: Vec::from(&b"beta"[..]).into(),
        }),
    ]);

    let result = run_session(
        runtime,
        b"",
        output,
        Box::pin(RecordingInputWriter::new()),
        host_stdout,
        host_stderr,
    );

    assert!(result.is_ok(), "protocol proxy should succeed: {result:?}");
    assert_eq!(
        stdout_bytes
            .lock()
            .expect("writer mutex should not poison")
            .clone(),
        b"alphabeta"
    );
    assert!(
        stderr_bytes
            .lock()
            .expect("writer mutex should not poison")
            .is_empty(),
        "stderr should remain untouched"
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=69&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fproxy_helpers.rs&method=)
The module contains 3 functions with similar structure: protocol_proxy_ignores_stdin_echo_chunks,protocol_proxy_routes_stderr_to_host_stderr,protocol_proxy_routes_stdout_and_console_to_host_stdout

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/engine/connection/exec/tests/protocol_proxy_bdd.rs](https://github.com/leynos/podbot/pull/69/files/ed7ec8a188ab254d83a447b0ed1d4c6629475307#diff-53118f4c052a3bcdbe04e6d93188ee187c8f2e743cbf34e81ba1721a29557d9e)

Comment on file

~~~rust
fn host_stdout_receives(protocol_proxy_state: &ProtocolProxyState, text: String) -> StepResult<()> {
    assert_channel_receives(
        &normalize_feature_text(&text),
        protocol_proxy_state.host_stdout.get(),
        "host stdout",
    )
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=69&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fprotocol_proxy_bdd.rs&method=)
The module contains 3 functions with similar structure: container_stdin_receives,host_stderr_receives,host_stdout_receives

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[src/engine/connection/exec/tests/proxy_helpers.rs](https://github.com/leynos/podbot/pull/69/files/232144560683055ad12f6094937bd69ffbe5095c#diff-43c74a15cc94f67e05066650495d263bd03804e7fd59b63faa9ebe3513a3f7dd)

Comment on file

~~~rust
fn protocol_proxy_routes_stdout_and_console_to_host_stdout(runtime: RuntimeFixture) {
    let output = make_output_stream(vec![
        Ok(LogOutput::StdOut {
            message: Vec::from(&b"alpha"[..]).into(),
        }),
        Ok(LogOutput::Console {
            message: Vec::from(&b"beta"[..]).into(),
        }),
    ]);

    let (result, stdout_bytes, stderr_bytes) = run_routing_session(runtime, output);

    assert!(result.is_ok(), "protocol proxy should succeed: {result:?}");
    assert_eq!(
        stdout_bytes
            .lock()
            .expect("writer mutex should not poison")
            .clone(),
        b"alphabeta"
    );
    assert!(
        stderr_bytes
            .lock()
            .expect("writer mutex should not poison")
            .is_empty(),
        "stderr should remain untouched"
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=69&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fproxy_helpers.rs&method=)
The module contains 2 functions with similar structure: protocol_proxy_routes_stderr_to_host_stderr,protocol_proxy_routes_stdout_and_console_to_host_stdout

File: src/engine/connection/exec/tests/proxy_helpers.rs [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+#[rstest]
+fn protocol_proxy_maps_daemon_stream_errors(runtime: RuntimeFixture) {
+    let output = make_output_stream(vec![Err(BollardError::RequestTimeoutError)]);
+
+    let result = run_session(
+        runtime,
+        b"",
+        output,
+        Box::pin(RecordingInputWriter::new()),
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/tests/proxy_helpers.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/engine/connection/exec/protocol.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+async fn settle_stdin_forwarding_task(
+    container_id: &str,
+    mut stdin_task: JoinHandle<io::Result<()>>,
+) -> Result<(), PodbotError> {
+    let Ok(join_result) = timeout(Duration::from_millis(50), &mut stdin_task).await else {
+        // The container output path has already completed, so stdin can be
+        // cancelled instead of waiting indefinitely on a live host reader. A
+        // short grace period above still captures EOF- and flush-driven
+        // completion paths deterministically.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/protocol.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The 50ms timeout in `settle_stdin_forwarding_task` is a hard-coded magic value; consider promoting this to a named constant (or configuration point) with rationale so future maintainers can tune it without digging into the implementation.
- The test helpers for proxy IO (e.g. `RecordingWriter` / `RecordingInputWriter` in `proxy_helpers.rs` and `ScenarioWriter` / `ScenarioInputWriter` in `protocol_proxy_bdd.rs`) duplicate similar behaviour; consider consolidating these into shared test utilities to reduce repetition and keep behaviour aligned.

## Individual Comments

### Comment 1
<location path="src/engine/connection/exec/tests/proxy_helpers.rs" line_range="212-221" />
<code_context>
+fn protocol_proxy_forwards_stdin_bytes_and_shutdowns_input(runtime: RuntimeFixture) {
</code_context>
<issue_to_address>
**suggestion (testing):** Consider adding a test for the case where host stdin never reaches EOF to validate non-hanging behaviour

You already cover the EOF case well. Please also add a test where host stdin never reaches EOF (e.g. a `DuplexStream` with the write end kept alive in a background task) to confirm that `settle_stdin_forwarding_task` still terminates via timeout/abort and returns `Ok(())` rather than hanging.

Suggested implementation:

```rust
#[rstest]
fn protocol_proxy_forwards_stdin_bytes_and_shutdowns_input(runtime: RuntimeFixture) {

```

You’ll need to add a new test function in `src/engine/connection/exec/tests/proxy_helpers.rs` (right after `protocol_proxy_forwards_stdin_bytes_and_shutdowns_input` is a good spot). The test should look roughly like this, adapted to your existing helpers and types:

```rust
#[rstest]
fn protocol_proxy_non_eof_stdin_does_not_hang(runtime: RuntimeFixture) {
    use tokio::io::duplex;

    // Arrange: create a duplex stream to simulate host stdin
    let (mut stdin_write, stdin_read) = duplex(1024);

    // Keep the write half alive in a background task so EOF is never reached
    runtime.spawn(async move {
        // Optionally write some bytes, but do NOT drop `stdin_write`
        let _ = stdin_write.write_all(b"partial input").await;
        // Then just park forever; the important part is that the handle stays alive
        futures::future::pending::<()>().await;
    });

    let host_stdout = RecordingWriter::new();
    let host_stderr = RecordingWriter::new();
    let container_input = RecordingInputWriter::new();
    let expected_shutdown = container_input.shutdown_called.clone();
    let output = make_output_stream(vec![]);

    // Act: run the session (or call `settle_stdin_forwarding_task` directly,
    // depending on how your tests are structured)
    let result = run_session(
        runtime,
        stdin_read,                            // host stdin that never reaches EOF
        output,
        Box::pin(container_input),
        host_stdout,
        host_stderr,
        /* any timeout / options your real signature takes */,
    );

    // Assert: the forwarding task is settled (did not hang) and we get Ok(())
    assert!(result.is_ok(), "session with non-EOF stdin should not hang");

    // We also expect container input to be shutdown even though stdin never hit EOF.
    assert!(
        expected_shutdown.load(Ordering::SeqCst),
        "container input should be shutdown even when host stdin never reaches EOF"
    );
}
```

Adjustments you will likely need to make:

1. **Import / runtime**  
   - If `tokio::io::duplex` is not already imported, add:
     ```rust
     use tokio::io::duplex;
     ```
   - Use your existing runtime-execution helper (e.g. `runtime.block_on` or similar) if `run_session` is async. In many of your tests, you may already wrap calls like:
     ```rust
     let result = runtime.block_on(run_session(...));
     ```

2. **`run_session` signature**  
   - Adapt the call to `run_session` to match its real parameters. In your current EOF test, mirror whatever you pass there (host stdout/stderr, output stream, options).
   - If `settle_stdin_forwarding_task` is not implicitly exercised by `run_session`, instead call it directly, e.g.:
     ```rust
     let result = runtime.block_on(settle_stdin_forwarding_task(
         stdin_read,
         Box::pin(container_input),
         timeout_duration,
     ));
     ```

3. **Timeout / abort configuration**  
   - If the non-hanging behaviour is controlled by a timeout or abort handle, make sure this test sets that timeout to a reasonably short value (e.g. a few hundred ms) so the test completes quickly.
   - Then assert that the returned `Result` is `Ok(())` even when it terminates via timeout/abort.

4. **Assertions**  
   - If your existing tests already assert on `stdout` / `stderr` contents, you can omit those here; the focus of this test is “no hang + Ok(()) + shutdown invoked”. Keep the assertions consistent with the behaviour of your `RecordingInputWriter` and the rest of the test suite.

By wiring the test as above, you validate that the stdin forwarding logic correctly terminates via timeout/abort and doesn’t rely solely on seeing EOF on host stdin.
</issue_to_address>

### Comment 2
<location path="src/engine/connection/exec/protocol.rs" line_range="106" />
<code_context>
+        return Ok(());
+    };
+
+    match join_result {
+        Ok(Ok(())) => Ok(()),
+        Ok(Err(error)) => Err(exec_failed(
</code_context>
<issue_to_address>
**issue (review_instructions):** This match expression has more than two branches and should be extracted into a dedicated predicate/helper function per the guideline.

The `match join_result` here has four branches (`Ok(Ok)`, `Ok(Err)`, `Err(cancelled)`, `Err(other)`). To comply with the "Move conditionals with >2 branches into a predicate function" rule, consider extracting this into a small helper (e.g. `fn classify_stdin_task_result(...) -> Result<(), PodbotError>`), and have `settle_stdin_forwarding_task` delegate to it.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>

### Comment 3
<location path="src/engine/connection/exec/protocol.rs" line_range="153" />
<code_context>
+        let chunk = chunk_result
+            .map_err(|error| exec_failed(container_id, format!("exec stream failed: {error}")))?;
+
+        match chunk {
+            LogOutput::StdOut { message } | LogOutput::Console { message } => {
+                write_output_chunk(container_id, host_stdout, message.as_ref(), "stdout").await?;
</code_context>
<issue_to_address>
**issue (review_instructions):** This multi-branch match over `LogOutput` should be moved into a helper/predicate function to satisfy the conditional branching guideline.

`run_output_loop_async` contains a `match chunk` with three logical branches (`StdOut/Console`, `StdErr`, and `StdIn`). To align with the "Move conditionals with >2 branches into a predicate function" instruction, you could extract this into a helper such as `fn handle_log_output(...)` and keep the loop focused on iteration and error handling.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

File: src/engine/connection/exec/tests/proxy_helpers/forwarding.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+use bollard::container::LogOutput;
+use rstest::rstest;
+use tokio::io::{AsyncWriteExt, duplex};
+
+use super::super::super::protocol::{ProtocolProxyIo, run_protocol_session_with_io_async};
+use super::*;
+
+#[rstest]
+fn protocol_proxy_forwards_stdin_bytes_and_shutdowns_input(runtime: RuntimeFixture) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/tests/proxy_helpers/forwarding.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
